src/main/java/hudson/plugins/ansicolor/AnsiHtmlOutputStream.java

```diff
@@ -93,12 +94,19 @@ private void stopConcealing() {
         this.out = logOutput;
     }
 
-    private void openTag(AnsiAttributeElement tag) throws IOException {
+    /**
+     * @return A copy of the {@link AnsiAttributeElement}s which are currently opened, in order from outermost to innermost tag.
+     */
+    /*package*/ List<AnsiAttributeElement> getOpenTags() {
+        return new ArrayList<>(openTags);
+    }
+
+    /*package*/ void openTag(AnsiAttributeElement tag) throws IOException {
```

Comment:
Making these methods accessible seems bad. Maybe better to provide a constructor that takes tags to open and opens them as the first part of the state machine, and then just `close` the stream in `ColorConsoleAnnotator` once we are done with it so `closeOpenTags` doesn't need to be called externally. I will investigate.

Reply:
It was easy to call `close` directly which removed the need to call `closeOpenTags`, but `openTag` is trickier. 

If the tags can be passed to the constructor, and are opened at the beginning of `AnsiHtmlOutputStream#write`, then in `ColorConsoleOutputStream` we end up with an `IndexOutOfBoundsException` and malformed output because the `CountingInputStream` records the first byte written before the initial tags are opened, so the offsets are incorrect. Calling `openTags` directly keeps the `CountingInputStream` from updating, which keeps this from being a problem. However, I think this means that there is already a bug in the existing implementation for any color map that sets a default foreground or background, since [this call](https://github.com/jenkinsci/ansicolor-plugin/blob/39f63c28386f432c4ff17ea4de8f4f2ba21a2089/src/main/java/hudson/plugins/ansicolor/AnsiHtmlOutputStream.java#L205) in `AnsiHtmlOutputStream#write` will cause the first character in the output to be hidden incorrectly.

I will keep thinking about the best way to handle this.

Reply:
Thinking about this some more, since `AnsiAttributeElement` is already package-private, I don't think this is a big problem, but either way I switched to using the constructor for open tags in `2c19b61` so that I could also fix the issue with default foreground/background colors.